src/paths/shortest_paths.c

```diff
@@ -1231,26 +1255,35 @@ int igraph_diameter_dijkstra(const igraph_t *graph,
     if (pto) {
         *pto = (igraph_integer_t) to;
     }
-    if (path) {
+    if ((vertex_path) || (edge_path)) {
         if (!igraph_finite(res)) {
-            igraph_vector_clear(path);
+            if (vertex_path){
+                igraph_vector_clear(vertex_path);
+            }
+            if (edge_path) {
+                igraph_vector_clear(edge_path);
+            }
         } else {
-            igraph_vector_ptr_t tmpptr;
-            igraph_vector_ptr_init(&tmpptr, 1);
-            IGRAPH_FINALLY(igraph_vector_ptr_destroy, &tmpptr);
-            VECTOR(tmpptr)[0] = path;
-            IGRAPH_CHECK(igraph_get_shortest_paths_dijkstra(graph,
-                         /*vertices=*/ &tmpptr, /*edges=*/ 0,
-                         (igraph_integer_t) from,
-                         igraph_vss_1((igraph_integer_t) to),
-                         weights, dirmode, /*predecessors=*/ 0,
-                         /*inbound_edges=*/ 0));
-            igraph_vector_ptr_destroy(&tmpptr);
-            IGRAPH_FINALLY_CLEAN(1);
+            if ((vertex_path) && (edge_path)) {
+                IGRAPH_CHECK(igraph_get_shortest_path_dijkstra(graph,
+                            /*vertices=*/ vertex_path, /*edges=*/ edge_path,
+                            (igraph_integer_t) from,
+                            (igraph_integer_t) to,
+                            weights, dirmode));
+            } else if (vertex_path) {
+                IGRAPH_CHECK(igraph_get_shortest_path_dijkstra(graph,
+                            /*vertices=*/ vertex_path, /*edges=*/ 0,
+                            (igraph_integer_t) from,
+                            (igraph_integer_t) to,
+                            weights, dirmode));
+            } else if (edge_path) {
+                IGRAPH_CHECK(igraph_get_shortest_path_dijkstra(graph,
+                            /*vertices=*/ 0, /*edges=*/ edge_path,
+                            (igraph_integer_t) from,
+                            (igraph_integer_t) to,
+                            weights, dirmode));
+            }
```

Comment:
It seems to me that we don't need if `if ... else` construct here. Just call `igraph_get_shortest_path` directly with `vertex_path` and `edge_path`.